src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs

```diff
@@ -125,5 +147,110 @@ public static partial class JsonSerializer
+        {
+            bool isPreserved = state.Current.IsProcessingProperty(ClassType.Dictionary) ?
+                state.Current.DictionaryPropertyIsPreserved : state.Current.IsPreserved;
+
+            if (metadata == MetadataPropertyName.Id)
+            {
+                // Check we are not parsing into immutable.
+                if (state.Current.TempDictionaryValues != null)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(state.Current.JsonPropertyInfo.DeclaredPropertyType);
+                }
+
+                if (state.Current.KeyName != null || isPreserved)
+                {
+                    state.Current.KeyName = null;
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                SetAsPreserved(ref state.Current);
+                state.Current.ReadMetadataValue = true;
+            }
+            else if (metadata == MetadataPropertyName.Ref)
+            {
+                if (state.Current.KeyName != null || isPreserved)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+                }
+
+                state.Current.ShouldHandleReference = true;
+                state.Current.ReadMetadataValue = true;
+            }
+        }
+
+        private static void ResolveMetadataOnObject(ReadOnlySpan<byte> propertyName, MetadataPropertyName meta, ref ReadStack state, ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            if (meta == MetadataPropertyName.NoMetadata)
+            {
+                if (state.Current.IsPreservedArray)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(state.Current.JsonClassInfo.PropertyCache["Values"].DeclaredPropertyType);
+                }
+                // Regular property, call main logic for HandlePropertyName.
+                HandlePropertyNameDefault(propertyName, ref state, ref reader, options);
+            }
+            else if (meta == MetadataPropertyName.Id)
+            {
+                if (state.Current.PropertyIndex > 0 || state.Current.IsPreserved)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataIdIsNotFirstProperty();
+                }
+
+                JsonPropertyInfo info = JsonPropertyInfo.s_metadataProperty;
+                info.JsonPropertyName = propertyName.ToArray();
+                state.Current.JsonPropertyInfo = info;
+
+                state.Current.ReadMetadataValue = true;
+                SetAsPreserved(ref state.Current);
+            }
+            else if (meta == MetadataPropertyName.Values)
+            {
+                // Preserved JSON arrays are wrapped into JsonPreservedReference<T> where T is the original type of the enumerable
```

Comment:
Shouldn't we validate that we are actually processing a non-dictionary collection? What if we had 

```c#
...
JsonSerializer.Deserialize<POCO>(@"{""$id"":""1"",""$values"":[1,2]}", options);
```

How could we tell (since we have coerced the class type to object using `JsonPreservedReference<T>`)?

Reply:
>Shouldn't we validate that we are actually processing a non-dictionary collection? 

I don't see the need; we check up-front that we are handling a JSON array (using `IsProcessingEnumerable()`), either on [HandleStartObject](https://github.com/dotnet/runtime/pull/655/files#diff-7d074ddc1c16d67315f2399ce9e1c2a6R27) and [HandleStartDictionary](https://github.com/dotnet/runtime/pull/655/files#diff-3098dbea2ba1872b07a0eb2ec5237f77R56) before wrapping it into a `JsonPreservedReference<T>`.